plugin/storage/badger/factory.go

```diff
@@ -0,0 +1,179 @@
+	}
+	if LastMaintenanceRun == nil {
+		LastMaintenanceRun = expvar.NewInt("badger_storage_maintenance_last_run")
+	}
+	if LastValueLogCleaned == nil {
+		LastValueLogCleaned = expvar.NewInt("badger_storage_valueloggc_last_run")
+	}
+	return &Factory{
+		Options:             NewOptions("badger"),
+		maintenanceInterval: defaultTickerInterval,
+	}
+}
+
+// AddFlags implements plugin.Configurable
+func (f *Factory) AddFlags(flagSet *flag.FlagSet) {
+	f.Options.AddFlags(flagSet)
+}
+
+// InitFromViper implements plugin.Configurable
+func (f *Factory) InitFromViper(v *viper.Viper) {
+	f.Options.InitFromViper(v)
+}
+
+// Initialize implements storage.Factory
+func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger) error {
+	f.logger = logger
+
+	opts := badger.DefaultOptions
+
+	if f.Options.primary.Ephemeral {
+		opts.SyncWrites = false
+		// Error from TempDir is ignored to satisfy Codecov
+		dir, _ := ioutil.TempDir("", "badger")
+		f.tmpDir = dir
+		opts.Dir = f.tmpDir
+		opts.ValueDir = f.tmpDir
+
+		f.Options.primary.KeyDirectory = f.tmpDir
+		f.Options.primary.ValueDirectory = f.tmpDir
+	} else {
+		opts.SyncWrites = f.Options.primary.SyncWrites
+		opts.Dir = f.Options.primary.KeyDirectory
+		opts.ValueDir = f.Options.primary.ValueDirectory
+	}
+
+	store, err := badger.Open(opts)
+	if err != nil {
+		return err
+	}
+	f.store = store
+
+	// Error ignored to satisfy Codecov
+	cache, _ := badgerStore.NewCacheStore(f.store, f.Options.primary.SpanStoreTTL)
+	f.cache = cache
+
+	f.maintenanceTicker = time.NewTicker(f.maintenanceInterval)
+	go f.maintenance()
+
+	logger.Info("Badger storage configuration", zap.Any("configuration", opts))
+
+	return nil
+}
+
+// CreateSpanReader implements storage.Factory
+func (f *Factory) CreateSpanReader() (spanstore.Reader, error) {
+	return badgerStore.NewTraceReader(f.store, f.cache), nil
+}
+
+// CreateSpanWriter implements storage.Factory
+func (f *Factory) CreateSpanWriter() (spanstore.Writer, error) {
+	return badgerStore.NewSpanWriter(f.store, f.cache, f.Options.primary.SpanStoreTTL, f), nil
+}
+
+// CreateDependencyReader implements storage.Factory
+func (f *Factory) CreateDependencyReader() (dependencystore.Reader, error) {
+	return nil, nil
+}
+
+// Close Implements io.Closer and closes the underlying storage
+func (f *Factory) Close() error {
+	f.maintenanceTicker.Stop()
+
+	err := f.store.Close()
+	if err != nil {
+		return err
+	}
+
+	// Remove tmp files if this was ephemeral storage
+	if f.Options.primary.Ephemeral {
+		err = os.RemoveAll(f.tmpDir)
+	}
+
+	return err
+}
+
+// Maintenance starts a background maintenance job for the badger K/V store, such as ValueLogGC
+func (f *Factory) maintenance() {
+	_, previousSize := f.store.Size()
+	for range f.maintenanceTicker.C {
+		_, vlogSize := f.store.Size()
+		if (previousSize + 1<<30) < vlogSize {
+			previousSize = vlogSize
+			var err error
+			for err == nil {
+				err = f.store.RunValueLogGC(0.5)
```

Comment:
what is `0.5`?

Reply:
It's a discard ratio, as in if half of the file can be reclaimed it will be rewritten. The 0.5 was the recommendation/sane default in badger's code: https://github.com/dgraph-io/badger/blob/fbb27786246d20213f31bad8bf33f1504837029f/db.go#L988

All the values are compromises so I chose this as a safe one.